ironfish/src/syncer.ts

```diff
@@ -99,46 +111,171 @@ export class Syncer {
+          break
+        }
+
+        this.logger.info('Checking for a potentially better peer to sync from')
+
+        // If it is time to enter the measuring state, we stop syncing and set
+        // the state to measuring. This gives the syncer time to finish any
+        // in-flight requests
+        this.state = 'measuring'
+        if (this.loader) {
+          this.lastLoaderIdentity = this.loader.getIdentity()
+          await this.wait()
+        }
+        break
+      }
     }
 
-    this.eventLoopTimeout = setTimeout(() => this.eventLoop(), SYNCER_TICK_MS)
+    this.eventLoopTimeout = setTimeout(() => void this.eventLoop(), SYNCER_TICK_MS)
   }
 
-  findPeer(): void {
+  /**
+   * Chooses a peer to sync from based on measuring the connection of a random
+   * sampling of connected peers and begins syncing
+   */
+  async findPeer(currentPeerIdentity: Identity | null): Promise<void> {
     const head = this.chain.head
-
     if (!head) {
       return
     }
 
+    this.logger.debug('Syncer is beginning peer candidate measurements')
+    const measurementStart = BenchUtils.start()
+
     // Find all allowed peers that have more work than we have
     const peers = this.peerNetwork.peerManager
       .getConnectedPeers()
       .filter((peer) => peer.features?.syncing && peer.work && peer.work > head.work)
 
-    // Get a random peer with higher work. We do this to encourage
-    // peer diversity so the highest work peer isn't overwhelmed
-    // as well as helping to make sure we don't get stuck with unstable peers
-    if (peers.length > 0) {
-      const peer = ArrayUtils.sampleOrThrow(peers)
+    if (peers.length === 0) {
+      return
+    }
+
+    // If there is only one valid peer to sync from, there is no point in
+    // measuring the connection so begin syncing immediately
+    if (peers.length === 1) {
+      this.startSync(peers[0])
+      return
+    }
+
+    const syncCandidates = ArrayUtils.shuffle(peers)
+
+    // If we have been syncing from a peer, we want to include this peer in the
+    // measurement. This will allow us to maintain a connection to a strong peer
+    // if we have found one.
+    if (
+      currentPeerIdentity &&
+      !syncCandidates.find((p) => p.getIdentity() === currentPeerIdentity)
+    ) {
+      const currentPeer = this.peerNetwork.peerManager.getPeer(currentPeerIdentity)
+      if (currentPeer && currentPeer.state.type === 'CONNECTED') {
+        syncCandidates.push(currentPeer)
+      }
+    }
+
+    const peerRtt = new Map<Identity, number>()
+
+    // Measure how long it takes to fetch the genesis block header from each
+    // peer as an estimate of connection quality
+    let candidateCheckCount = 0
+    for (const peer of syncCandidates) {
+      if (this.state === 'stopped' || this.state === 'stopping') {
+        return
+      }
+
+      // We only want to successfully measure so many candidates per measurement
+      // phase
+      if (candidateCheckCount > 8) {
```

Comment:
You could potentially use `peerRtt.size` for this instead. I think this will allow 9 now where the previous algo allowed 8 (+1 if currentPeerIdentity was set), but I don't think the exact number is too important. Maybe it could be pulled out to a constant too?